widgetssdk/src/main/java/com/glia/widgets/chat/ChatView.java

```diff
@@ -1073,15 +1135,121 @@ private static Uri chooseUriByRequestCode(int requestCode, Uri galeryImgUri, Uri
     public void onActivityResult(int requestCode, int resultCode, Intent intent) {
         if ((requestCode == OPEN_DOCUMENT_ACTION_REQUEST || requestCode == CAPTURE_IMAGE_ACTION_REQUEST || requestCode == CAPTURE_VIDEO_ACTION_REQUEST)
                 && resultCode == Activity.RESULT_OK) {
-
-            Uri uri = chooseUriByRequestCode(requestCode, intent.getData(), photoCaptureFileUri);
-
+            Uri dataUri = intent != null ? intent.getData() : null;
+            Uri uri = chooseUriByRequestCode(requestCode, dataUri, photoCaptureFileUri);
             if (uri != null) {
                 controller.onAttachmentReceived(uri);
             }
         }
     }
 
+    @Override
+    public void onFileDownloadClick(OperatorAttachmentItem clickedItem) {
+        submitUpdatedItems(clickedItem, true, false);
+
+        Context context = this.getContext();
+        AttachmentFile clickedFile = clickedItem.attachmentFile;
+        File file = new File(context.getFilesDir(), clickedFile.getName());
+        downloadFile(clickedItem, file);
+    }
+
+    private void downloadFile(OperatorAttachmentItem clickedItem, File file) {
+        if (clickedItem.attachmentFile.isDeleted()) {
+            Toast.makeText(this.getContext(), this.getContext().getString(R.string.chat_file_download_failed_msg), Toast.LENGTH_SHORT).show();
+            return;
+        }
+
+        Glia.fetchFile(clickedItem.attachmentFile, (fileInputStream, gliaException) -> new Thread(() -> {
```

Comment:
there could be a FileProviderRepository made for that case - so that single object would be responsible for fetching the files

Reply:
tho I see that there could be problems with messageList atm. might need to sync later

Reply:
In this case, too many listeners need to be involved since the use case cannot be called from ChatView directly. I kinda do not want to do it since it just creates unnecessary complexity.

Reply:
it could react on the itemslist state change aswell invoked after files get fetched

but its not totally "stable" atm. so I see your concerns

Reply:
Totally, so agreed on keeping it as it is?